### PR TITLE
Add Pagination on the Apply Filter Button

### PR DESCRIPTION
**Pull request by Achintya-Chatterjee**
Date: 07 February 2024

Developer Name: @Achintya-Chatterjee 

---

## Issue Ticket Number
 

https://github.com/Real-Dev-Squad/website-dashboard/issues/678

## Description
Pagination is missing when multiple filter is applied and user count is 100+

### Documentation Updated?

- [x] Yes
- [ ] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

[Screencast from 09-02-24 02:15:52 AM IST.webm](https://github.com/Real-Dev-Squad/website-dashboard/assets/55826451/d463a651-a943-43da-8641-229c34ea7e83)

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

Screenshot from 2024-02-09 02-07-12
Screenshot from 2024-02-13 13-24-14
Screenshot from 2024-02-13 13-27-17

</details>

### CODE REVIEW

File: __tests__/users/applyFilterPagination.test.js
Comment (sahsisunny) [inline __tests__/users/applyFilterPagination.test.js:1]: Is this test is unit or integration?

File: __tests__/users/App.test.js [hunk 1]
```diff
@@ -49,6 +50,8 @@ describe('App Component', () => {
           body: JSON.stringify({
             ...filteredUsersData,
             users: filteredUsersData.users.filter((user) => user.discordId),
+            ...mockUserData,
+            users: mockUserData.users.filter((user) => user.discordId),
```
Comment (sahsisunny) [inline __tests__/users/App.test.js:54]: Why are you filtering users by discord ID?

**Review by sahsisunny — CHANGES_REQUESTED** [review]
Could you please provide tests stats for your changes? If not available, kindly specify the title of the test that passed during testing.

**Comment by ajoykumardas12** [comment]
Is there next or previous pages of pagination for the visual provided? If not the buttons should be in disabled state.
Also, can you please provide a demo where visiting next, previous pages work? (Can be done by decreasing users limit temporarily  for testing purpose I think).

File: users/script.js
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply users/script.js:1]: remove commented code, forgot to remove it when committing the changes to GitHub

File: mock-data/users/mockdata.js
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply mock-data/users/mockdata.js:1]: I was trying to use those mock-data, but it gave me so many errors. That's why added a new mock-data for testing of Apply Filter Pagination

**Comment by Achintya-Chatterjee** [comment]
> Could you please provide tests stats for your changes? If not available, kindly specify the title of the test that passed during testing.

@sahsisunny already provided tests stats for my changes

**Comment by Achintya-Chatterjee** [comment]
> Is there next or previous pages of pagination available for the visual provided? If not the buttons should be in disabled state. Also, can you please provide a demo where visiting next, previous pages work? (Can be done by decreasing users limit temporarily for testing purpose I think).

@ajoykumardas12 cool, I will provide a video

**Comment by sahsisunny** [comment]
> > Could you please provide tests stats for your changes? If not available, kindly specify the title of the test that passed during testing.
> 
> @sahsisunny already provided tests stats for my changes

Where?

File: users/script.js [hunk 2]
```diff
@@ -328,22 +329,24 @@ function clearFilters() {
...
   prevBtn,
   nextBtn,
-) {
+) => {
   try {
+    // Fetch users data with pagination parameters
     const userData = await getUsersData(page);
     if (userData.length) {
-      if (userData.length < 100) {
+      if (userData.length < USER_FETCH_COUNT) {
```
Comment (vinit717) [inline users/script.js:344]: where this defined USER_FETCH_COUNT?

File: users/script.js [hunk 3]
```diff
@@ -328,22 +329,24 @@ function clearFilters() {
   );
 }
 
-async function showUserDataList(
+const showUserDataList = async (
```
Comment (vinit717) [inline users/script.js:332]: Why you are changing function declaration to arrow function?

File: users/script.js
Comment (vinit717) [inline users/script.js:1]: Can you please go through the comments you have added many places I don't think there is a need of this
and also go to the code once can you simply it

Comment (Achintya-Chatterjee) [reply users/script.js:1]: Removed all of those comments. I just added those comments for my understanding

File: __tests__/users/applyFilterPagination.test.js
Comment (Achintya-Chatterjee) [reply __tests__/users/applyFilterPagination.test.js:1]: `applyFilterPagination.test.js` is an integration test as it tests the interaction between different components of the application. It primarily interacts with the UI, it indirectly tests backend functionality related to URL manipulation.

File: __tests__/users/App.test.js (hunk 1 above)
Comment (Achintya-Chatterjee) [reply __tests__/users/App.test.js:54]: filtering of users by Discord ID in the test is intentional to ensure that our application accurately handles and displays data specific to users with Discord IDs. This approach allows us to comprehensively test our application's functionality. If you have any further questions or suggestions, please let me know.

File: users/script.js (hunk 2 above)
Comment (Achintya-Chatterjee) [reply users/script.js]: `USER_FETCH_COUNT` constant is defined in another file where constants are typically declared. It represents the maximum number of users to fetch per request.

File: users/script.js (hunk 3 above)
Comment (Achintya-Chatterjee) [reply users/script.js:331]: The change in function declaration from function expression to arrow function was made to maintain consistency in coding style and improve readability. Arrow functions provide a more concise syntax and lexical scoping, which can be beneficial in certain contexts. If you prefer the original function declaration, I can revert it.

**Review by palakgupta2712 — DISMISSED** [review]

**Review by sahsisunny — DISMISSED** [review]

File: users/script.js (hunk 2 above)
Comment (Achintya-Chatterjee) [reply users/script.js]: https://github.com/Real-Dev-Squad/website-dashboard/blob/develop/users/constants.js @vinit717 you can see this file for references.

File: users/script.js [hunk 4]
```diff
@@ -588,45 +588,48 @@ applyFilterButton.addEventListener('click', async () => {
...
+      users,
+      true,
+      userListElement,
+      paginationElement,
+      loaderElement,
+      prevBtn,
+    );
   } catch (err) {
-    throw new Error(`User list request failed with error: ${err}`);
+    console.error(`User list request failed with error: ${err}`);
```
Comment (yesyash) [inline users/script.js:626]: - use `console.debug`

Comment (Achintya-Chatterjee) [reply users/script.js]: Done. Remove console log and added console debug @yesyash

Comment (sahsisunny) [reply users/script.js]: we can use console.error instead of debug right @yesyash ?

Comment (yesyash) [reply users/script.js]: In the error above i suggested to use console.debug as the error was not valid to be shown to user in console, if we we were logging all console errors for future reference i would have suggested console error

**Review by yesyash — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Pull request merged**